### ISSUE TITLE
Auto-correct does not cache

### ISSUE THREAD

**Issue by cabello** [bbatsov/rubocop#5082]
The first time I run `bundle exec rubocop` is slower than the sub-sequent ones because of the awesome caching feature.

When using the auto-correct flag `-a`, it seems to never cache and every single run takes on average the same amount of time.

## Expected behavior

Caching should also work when auto-correcting files.

## Actual behavior

Every run takes consistently the same amount of time.

## Steps to reproduce the problem

```
➤ time bundle exec rubocop -a
Inspecting 269 files
.............................................................................................................................................................................................................................................................................

269 files inspected, no offenses detected
       10.30 real         9.93 user         0.29 sys
➤ time bundle exec rubocop -a
Inspecting 269 files
.............................................................................................................................................................................................................................................................................

269 files inspected, no offenses detected
       10.18 real         9.80 user         0.29 sys
➤ time bundle exec rubocop -a
Inspecting 269 files
.............................................................................................................................................................................................................................................................................

269 files inspected, no offenses detected
       10.16 real         9.77 user         0.31 sys
```

Compare to:

```
➤ time bundle exec rubocop
Inspecting 269 files
.............................................................................................................................................................................................................................................................................

269 files inspected, no offenses detected
       10.99 real        10.54 user         0.40 sys
➤ time bundle exec rubocop
Inspecting 269 files
.............................................................................................................................................................................................................................................................................

269 files inspected, no offenses detected
        1.31 real         1.04 user         0.25 sys
```

## RuboCop version

```
➤ bundle exec rubocop -V
0.49.1 (using Parser 2.4.0.0, running on ruby 2.3.5 x86_64-darwin16)
```

Is that a reasonable request? Thanks for the awesome project! ❤️

**Comment by pocke** [comment]
It seems good to me. But we must refactor the code for auto-correction to implement this feature.

First, we should add `auto_correct` to `NON_CHANGING`.

```diff
diff --git a/lib/rubocop/result_cache.rb b/lib/rubocop/result_cache.rb
index bd5bb85e..01c240ef 100644
--- a/lib/rubocop/result_cache.rb
+++ b/lib/rubocop/result_cache.rb
@@ -7,7 +7,7 @@ require 'etc'
 module RuboCop
   # Provides functionality for caching rubocop runs.
   class ResultCache
-    NON_CHANGING = %i[color format formatters out debug fail_level
+    NON_CHANGING = %i[color format formatters out debug fail_level auto_correct
                       cache fail_fast stdin parallel].freeze
 
     # Remove old files so that the cache doesn't grow too big. When the
diff --git a/lib/rubocop/runner.rb b/lib/rubocop/runner.rb
index fc405261..46506c7c 100644
--- a/lib/rubocop/runner.rb
+++ b/lib/rubocop/runner.rb
@@ -175,8 +175,6 @@ module RuboCop
         # cops related to calculating the Max parameters for Metrics cops. We
         # need to do that processing and can not use caching.
         !@options[:auto_gen_config] &&
-        # Auto-correction needs a full run. It can not use cached results.
-        !@options[:auto_correct] &&
         # We can't cache results from code which is piped in to stdin
         !@options[:stdin]
     end
```

But it does not work well from we just changing them.
For example

```shell
$ cat test.rb
p "a"



$ rubocop 
Inspecting 1 file
C

Offenses:

test.rb:1:3: C: Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols.
p "a"
  ^^^

1 file inspected, 1 offense detected



# It does not execute auto-correction!
$ rubocop -a
Inspecting 1 file
C

Offenses:

test.rb:1:3: C: Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols.
p "a"
  ^^^

1 file inspected, 1 offense detected
```

**Comment by jonas054** [comment]
Thanks, @pocke for the initial work on this!
